rust/xaynet-analytics/src/controller.rs

```diff
@@ -148,3 +176,378 @@ impl AnalyticsController {
+        let name = "test";
+        let event_type = AnalyticsEventType::ScreenEnter;
+        let timestamp = DateTime::parse_from_rfc3339("2021-01-01T01:01:00+00:00")
+            .unwrap()
+            .with_timezone(&Utc);
+        let screen_route_name = "route";
+
+        assert!(controller
+            .save_analytics_event(name, event_type, timestamp, Some(screen_route_name))
+            .is_ok());
+
+        let screen_route = ScreenRoute::new(screen_route_name, timestamp);
+        let analytics_event = AnalyticsEvent::new(name, event_type, timestamp, Some(screen_route));
+        let all_analytics_events =
+            AnalyticsEvent::get_all(controller.db(), CollectionNames::ANALYTICS_EVENTS);
+        assert_eq!(
+            all_analytics_events.unwrap().first(),
+            Some(&analytics_event)
+        );
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_change_connectivity_status() {
+        let test_name = "test_change_connectivity_status";
+        let mut controller = get_controller(test_name, None);
+        assert_eq!(controller.is_connected_to_wifi, true);
+
+        controller.change_connectivity_status();
+        assert_eq!(controller.is_connected_to_wifi, false);
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_change_state_of_charge() {
+        let test_name = "test_change_state_of_charge";
+        let mut controller = get_controller(test_name, None);
+        assert_eq!(controller.is_charging, true);
+
+        controller.change_state_of_charge();
+        assert_eq!(controller.is_charging, false);
+
+        cleanup(controller, test_name);
+    }
+
+    #[test]
+    fn test_validate_send_data_frequency_when_none() {
+        assert_eq!(
+            AnalyticsController::validate_send_data_frequency(None).unwrap(),
+            Duration::hours(AnalyticsController::MAX_SEND_DATA_FREQUENCY_HOURS as i64)
+        )
+    }
+
+    #[test]
+    fn test_validate_send_data_frequency_when_more_than_24() {
+        assert!(AnalyticsController::validate_send_data_frequency(Some(25)).is_err());
+    }
+
+    #[test]
+    fn test_validate_send_data_frequency_when_less_than_24() {
+        assert_eq!(
+            AnalyticsController::validate_send_data_frequency(Some(6)).unwrap(),
+            Duration::hours(6)
+        )
+    }
+
+    #[test]
+    fn test_add_screen_route_if_new_with_new_route() {
+        let test_name = "test_add_screen_route_if_new_with_new_route";
+        let controller = get_controller(test_name, None);
+        let screen_route_name = "route";
+        let timestamp = DateTime::parse_from_rfc3339("2021-01-01T01:01:00+00:00")
+            .unwrap()
+            .with_timezone(&Utc);
+        let screen_route = ScreenRoute::new(screen_route_name, timestamp);
+        assert_eq!(
+            controller
+                .add_screen_route_if_new(screen_route_name, timestamp)
+                .unwrap(),
+            screen_route
+        );
```

Comment:
Not sure i understand, i think it's the next test, maybe i need to reword it or add some comment to make it more clear?

Reply:
Kind of, to test it in the next test we have to generate a new `controller` to use to insert the second screen route otherwise the controller could keep the screen routes in an internal state and never persist them on db.
Moreover in the next test we need to check that we have only one screen route on the database because `add_screen_route_if_new` could add every screen route to the db and just return the one with the lowest timestamp.

Reply:
ok got it. good points, will add it